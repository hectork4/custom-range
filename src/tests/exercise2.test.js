import { render, screen } from '@testing-library/react';
import Exercise2 from '../app/exercise2/page';
import { redirect } from 'next/navigation';
import {
  NO_VALUES_MESSAGE,
  FAILED_FETCH_MESSAGE,
} from '@/constants';

jest.mock('next/navigation', () => ({
  redirect: jest.fn(),
}));

global.fetch = jest.fn();

describe('Exercise2', () => {
  let consoleErrorSpy;

  beforeEach(() => {
    fetch.mockClear();
    redirect.mockClear();
    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    consoleErrorSpy.mockRestore();
  });

  it('renders correctly with valid data', async () => {
    fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({ rangeValues: [0, 50, 100] }),
    });

    const { container } = render(await Exercise2());

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Fixed Values')).toBeInTheDocument();
    expect(screen.getByText('0')).toBeInTheDocument();
    expect(screen.getByText('100')).toBeInTheDocument();
  });

  it('redirects when no data is available', async () => {
    fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({ rangeValues: [] }),
    });

    await Exercise2();

    expect(redirect).toHaveBeenCalledWith(`/?message=${NO_VALUES_MESSAGE}`);
  });

  it('redirects when data is invalid', async () => {
    fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({ rangeValues: null }),
    });

    await Exercise2();

    expect(redirect).toHaveBeenCalledWith(`/?message=${NO_VALUES_MESSAGE}`);
  });

  it('handles fetch error', async () => {
    fetch.mockRejectedValueOnce(new Error('Fetch error'));

    await Exercise2();

    expect(consoleErrorSpy).toHaveBeenCalledWith(FAILED_FETCH_MESSAGE, expect.any(Error));
    expect(redirect).toHaveBeenCalledWith(`/?message=${NO_VALUES_MESSAGE}`);
  });
});
