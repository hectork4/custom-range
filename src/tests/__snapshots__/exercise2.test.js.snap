// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercise2 renders correctly with valid data 1`] = `
<div>
  <div>
    <h1>
      Fixed Values
    </h1>
    <div
      class="range"
    >
      <div
        class="range-labels"
      >
        <div
          class="label-wrapper"
          style="left: -3%;"
        >
          <span>
            0
          </span>
        </div>
        <div
          class="label-wrapper"
          style="left: 100%;"
        >
          <span>
            100
          </span>
        </div>
      </div>
      <div
        class="range-slider"
      >
        <div
          aria-label="min"
          class="range-bullet min"
          role="slider"
          style="left: 0%;"
        />
        <div
          aria-label="max"
          class="range-bullet max"
          role="slider"
          style="left: 100%;"
        />
      </div>
    </div>
    <a
      class="btn btn-primary"
      href="/"
    >
      Go back
    </a>
  </div>
</div>
`;
